spec/index.bs

```diff
@@ -4183,41 +4189,71 @@ has a default layout created and used instead.
 
 ### <dfn dictionary>GPUProgrammableStage</dfn> ### {#GPUProgrammableStage}
 
+A {{GPUProgrammableStage}} describes the entry point in the user-provided
+{{GPUShaderModule}} that controls one of the programmable stages of a [=pipeline=].
+
 <script type=idl>
 dictionary GPUProgrammableStage {
     required GPUShaderModule module;
     required USVString entryPoint;
+    record<USVString, double> constants;
 };
 </script>
 
-A {{GPUProgrammableStage}} describes the entry point in the user-provided
-{{GPUShaderModule}} that controls one of the programmable stages of a [=pipeline=].
+<dl dfn-for=GPUProgrammableStage dfn-type=dict-member>
+    : <dfn>constants</dfn>
+    ::
+        Specifies the values of [=pipeline-overridable=] constants which are [=statically accessed=]
+        by the entry point {{GPUProgrammableStage/entryPoint}} in the shader module
+        {{GPUProgrammableStage/module}}.
 
-<div algorithm>
-    <dfn abstract-op>validating GPUProgrammableStage</dfn>(stage, descriptor, layout)
-        **Arguments:**
-            - {{GPUShaderStage}} |stage|
-            - {{GPUProgrammableStage}} |descriptor|
-            - {{GPUPipelineLayout}} |layout|
+        Each key-value pair must uniquely identify a [=pipeline-overridable=] constant
+        (by a string representing the numeric ID of the constant, if specified;
+        otherwise, by the constant's identifier name).
 
-        Return `true` if all of the following conditions are satisfied:
+        When the pipeline is executed, that constant will be set to the specified value
+        ([=converted to an IDL value|converted=] from `double` to
+            {{boolean}}, {{float}}, {{long}}, or {{unsigned long}},
+            depending on the [=wgsl-type=] of the [=wgsl-declaration=] in WGSL).
```

Comment:
I'm sure this was discussed and I simply wasn't at the WGSL meeting, but it seems strange on first read to consume all of these as doubles if we're going to be converting them internally. I guess `any` isn't appropriate here, but maybe a union? Point being, it would be nice to pass in booleans as booleans.

Reply:
We use a similar trick in GPUColor:

> Note: double is large enough to precisely hold 32-bit signed/unsigned integers and single-precision floats.

Though of course booleans are not allowed there, so this is different.

It's (presumably) not allowed to have unions of different number types (since how would you pick which union overload to use?), but I assume it is allowed to have `double or boolean`, so we could certainly do so here.

This raises the question though, what happens if you pass a boolean when a number was expected, or vice versa?
- We can't have a synchronous exception, because the client doesn't know what the shader interface is.
- We could have a validation error, but this would require implementations to keep track of the distinction between e.g. `false` and `0` through to the service, which adds some questionably-useful complexity.
- We could do implicit coercions between booleans and numbers, at which point we're doing exactly the same thing that IDL would do for us implicitly if we didn't use the union.

Reply:
Hm, I see. I guess we don't want to lift the reflection data up to the Blink level to validate against. 😕

> We could do implicit coercions between booleans and numbers, at which point we're doing exactly the same thing that IDL would do for us implicitly if we didn't use the union.

I would have thought that passing a bool to a double would throw a type error? If not and it coerces it to something like `1` and `0` then that's probably fine. Though in that case maybe we could use a typedef to clarify it's use? Something like:

```webidl
typedef double GPUPipelineConstant; // May be treated as a double, boolean, or integer
```

Reply:
> If not and it coerces it to something like `1` and `0` then that's probably fine.

Booleans coerce to 1 or +0, per WebIDL (which calls into the standard ECMAScript boolean-to-number conversion).